### ISSUE TITLE
Scratch 3's treatment of ±Infinity isn't the same as Scratch 2

### ISSUE THREAD

**Issue by TheLogFather** [LLK/scratch-vm#1699]
### Expected Behavior

In Scratch 2 the following expressions are all true:
< ( 1 / 0 ) = [Infinity] > where the right side is a string
< (var) = [Infinity] > when var was set to 1/0 (i.e. Infinity) and right side is a string
< [list v] contains [Infinity] > when it contains a numerical infinity (e.g. from 1/0)
< (var) = (var) > when var is ±Infinity (numerical, e.g. from ±1/0)
etc...

### Actual Behavior

In Scratch 3 they are all false.
The last one above is particularly unexpected.

### Steps to Reproduce

Test the expressions given above in Scratch 2 vs 3.

### Operating System and Browser

N/A

**Comment by Kenny2github** [comment]
To be mathematically accurate...
1/0 != Infinity. 1/0 is undefined, and should either be an exception (which wouldn't apply to Scratch) or NaN.
Numerical infinity does not equal infinity, as there can be multiple kinds of infinity. Consider the infinite series of natural numbers (i.e. 1, 2, 3...) vs the infinite series of even natural numbers (i.e. 2, 4, 6...). These are both infinity, but they are not equal. Therefore it is better to assume that infinity does not equal infinity in all cases.
<[list v] contains [Infinity]>, however, should report true if the list contains numerical infinity. It is safe to assume that the contains and the contained refer to the same infinity.
<(var) = (var)> should be false when either one is numerical infinity, but true when both are the string "Infinity".

Basically, except for the contains block, Scratch 3 is mathematically accurate... however, keeping it that way does present backwards compatibility issues. Therefore the decision here is between mathematical accuracy and backwards compatibility: which is more important?

**Comment by TheLogFather** [comment]
I don't think 'mathematical consideration' is really relevant to what I said above.

It was simply "the behaviour in 3 is inconsistent with 2".

It's worth noting that we're not quite talking about mathematics here. Some of this is really computational arithmetic (which is guided by mathematical principles and concepts, but it is a bit different), and some of it is how Scratch hides type from Scratchers (see below).

The question of how Scratch 'should' behave when it encounters potential 'mistakes' (including infinity/NaN/etc.) was discussed elsewhere some time back.
–My own take on it was that Scratch shouldn't stop with an error, but that it'd be good to develop a more general interface for all kinds of (potential) errors it encounters (including, for example, out-of-range list access, invalid costume, divide-by-zero, etc.) This means it would highlight the offending block in some way, and hovering over that block would provide a pop-up that explains what happened and what could be done to fix it.

Also, note that I am a mathematician, so I'm well aware of the curiosities of infinity. :)
(BTW, your assertion concerning the infinity of natural numbers vs even natural numbers isn't correct. Those two are the same infinity. The one you want is the 'uncountable infinity' of real numbers.)

**Comment by TheLogFather** [comment]
I think part of the issue here is that Scratch, most of the time, tries to do its best to hide the concept of 'type' from inexperienced coders. (The main place that didn't hold was for input boxes, where a round-end one was for numerical values, while square-end was more 'generic' – and clearly the hexagonal slot was for booleans, but you can't type into that.)

This 'philosophy' is generally handled very well in Scratch. In a few cases it causes trouble.

I think, in this case, following that 'philosophy' leads me to the conclusion that there is a good argument, that it's 'intuitively sensible', to have <(1/0)="Infinity"> be true in Scratch (where I've used the quotes to make it clear the RHS of the expression is a string).

Note that there are other related cases, such as <<1=1>="true"> and <<1=0>="false">.
Though there are others that are less obvious, such as <<1=1>=1>, and <<1=0>=0> (and these then lead to the oddity that 1=<1=1> and <1=1>="true", but clearly 1≠"true").

And these things all lead to the somewhat more tricky issue of [how number/boolean and string comparison works with localisation...](https://github.com/LLK/scratch-blocks/issues/1685)

**Comment by joker314** [comment]
Okay, I've figured out what's going on. Both Scratch 3.0 and Scratch 2.0 compare things by passing it `(a, b)` to a function:

- if the return value is negative, `a < b`
- if the return value is 0, `a = b`
- if the return value is positive, `a > b`

Notice that this bug only happens when you spell Infinity in a case-sensitive way
```
<(1 / 0) = "Infinity"> // false
<(1 / 0) = "infinity"> // true
```

The reason is that both versions will attempt to cast both arguments to numbers. And they do this using `Number(arg)`. This is case-sensitive. If it fails to convert the RHS (or LHS) to a number, it will compare as strings, and this bit of the code is backwards-compatible with 2.0.

**This bug only happens during numeric comparisons**. Let's take a look at how 2.0 does those:

```actionscript
			if (n1 < n2) return -1;
			if (n1 == n2) return 0;
			if (n1 > n2) return 1;
```

So if we compare `Infinity` and `Infinity`, we get a return value of 0.

However, this is how 3.0 does the numeric comparisons:
https://github.com/LLK/scratch-vm/blob/de86eb3f19cf812c37f03aff365c05cd41bd185f/src/util/cast.js#L129

If we try `Infinity - Infinity`, we get `NaN`, which isn't 0. Hence, they aren't considered equal. I'll try and send a PR later today to fix this.

**Comment by TheLogFather** [comment]
Ah, nice catch!

I hadn't tried "infinity", with lower-case "i", so that would certainly have been a useful indicator if I'd noticed it.

Also, I'd not noticed that this bug means even using strings on both sides would fail in Scratch 3  (because it ends up casting the string "Infinity" into a numerical value, as you showed above).

So <"Infinity" = "Infinity"> is false!

(Using lower-case "i" works fine, on either/both side(s).)

**Comment by thisandagain** [comment]
Oh boy. I remember dealing with this in Scratch 2.0. Thanks for filing this @TheLogFather.

**Comment by joker314** [comment]
(Click on the arrows to expand)

<details>
<summary>There are actually two differences in implementation between 2.0/3.0, and both of these changed slightly, which is why Infinity managed to have this bug. This section is just a slight correction regarding that</summary>
<hr>
My analysis was slightly out (both filed PRs still resolve the issue though). In Scratch 2.0, we make sure there's at least one digit in a number before treating it as such. In Scratch 3.0, we make sure there's at least one non-whitespace character. We assumed this would be the same thing, but... ±Infinity happened!

So Scratch 2.0 would always compare an infinity as if it were a string. This is why <Infinity = 10e100000000> comes out as false: the LHS has no digits in it.

However, even if Infinity were compared numerically in Scratch 2.0, it would still have worked, because (per my previous comments), in 2.0 we weren't subtracting the values, instead, we were comparing them.

Given that <em>both</em> of these things changed in Scratch 3.0, we had this issue happen.
</details>
<hr>

<details>
<summary>This issue does not happen with numbers other than +/-Infinity</summary>
<hr/>
I had a quick look at the specification, and it looks like this should be our only issue. Except for just whitespace, a string number must (save for leading/trailing whitespace) be:
<pre>
- StrDecimalLiteral
  - Which must contain a StrUnsignedDecimalLiteral
    - Which must contain DecimalDigit(s) or "Infinity"
- BinaryIntegerLiteral 
  - Which must contain (start with) a "0"
- OctalIntegerLiteral
  - Which must contain (start with) a "0"
- HexIntegerLiteral
  - Which must contain (start with) a "0"
